src/ccpp_constituent_prop_mod.F90

```diff
@@ -914,351 +1283,975 @@ subroutine ccp_model_const_lock(this, ncols, num_layers, num_interfaces,   &
+
+   subroutine ccpt_get_vertical_dimension(this, vert_dim, errcode, errmsg)
+      ! Return the standard name of this constituent's vertical dimension
+
+      ! Dummy arguments
+      class(ccpp_constituent_prop_ptr_t), intent(in)  :: this
+      character(len=*),                   intent(out) :: vert_dim
+      integer,          optional,         intent(out) :: errcode
+      character(len=*), optional,         intent(out) :: errmsg
+      ! Local variable
+      character(len=*), parameter :: subname = 'ccpt_get_vertical_dimension'
+
+      if (associated(this%prop)) then
+         if (this%prop%is_instantiated(errcode, errmsg)) then
+            vert_dim = this%prop%vert_dim
+         end if
+      else
+         vert_dim = ''
+         call set_errvars(1, subname//": invalid constituent pointer",        &
+              errcode=errcode, errmsg=errmsg)
+      end if
+
+   end subroutine ccpt_get_vertical_dimension
+
+   !#######################################################################
+
+   logical function ccpt_is_layer_var(this) result(is_layer)
+      ! Return .true. iff this constituent has a layer vertical dimension
+
+      ! Dummy arguments
+      class(ccpp_constituent_prop_ptr_t), intent(in)  :: this
+      ! Local variables
+      character(len=dimname_len)  :: dimname
+      character(len=*), parameter :: subname = 'ccpt_is_layer_var'
+
+      if (associated(this%prop)) then
+         call this%prop%vertical_dimension(dimname)
+         is_layer = trim(dimname) == 'vertical_layer_dimension'
+      else
+         is_layer = .false.
+      end if
+
+   end function ccpt_is_layer_var
+
+   !#######################################################################
+
+   logical function ccpt_is_interface_var(this) result(is_interface)
+      ! Return .true. iff this constituent has a interface vertical dimension
+
+      ! Dummy arguments
+      class(ccpp_constituent_prop_ptr_t), intent(in)  :: this
+      ! Local variables
+      character(len=dimname_len)  :: dimname
+      character(len=*), parameter :: subname = 'ccpt_is_interface_var'
+
+      if (associated(this%prop)) then
+         call this%prop%vertical_dimension(dimname)
+         is_interface = trim(dimname) == 'vertical_interface_dimension'
+      else
+         is_interface = .false.
+      end if
+
+   end function ccpt_is_interface_var
+
+   !#######################################################################
+
+   logical function ccpt_is_2d_var(this) result(is_2d)
+      ! Return .true. iff this constituent has a 2d vertical dimension
+
+      ! Dummy arguments
+      class(ccpp_constituent_prop_ptr_t), intent(in)  :: this
+      ! Local variables
+      character(len=dimname_len)  :: dimname
+      character(len=*), parameter :: subname = 'ccpt_is_2d_var'
+
+      if (associated(this%prop)) then
+         call this%prop%vertical_dimension(dimname)
+         is_2d = len_trim(dimname) == 0
```

Comment:
Not sure I understand the API of `vertical_dimension`.  Does it only put a result in `dimname` if the `vertical_dimension` sin't 2D?

Reply:
I'm not sure. @gold2718 might be able to answer better.

I think the vertical dimension is only populated (len > 0) if there is a vertical dimension (which makes the variable 3D)?

Reply:
For CAM at least, 2-D means there is _no_ vertical dimension so there is no name for it. It might be cleaner and more understandable if a `is_2d` method was added to the `prop` obect.

Reply:
@gold2718 this might not be where @mwaxmonsky 's confusion lies, but why is a variable without a vertical dimension considered "2d"? Wouldn't that be "1d"?

Reply:
Hmm. Okay, I'll try harder.
Because I said so :)

Or, because any physical quantity is going to be defined in every grid cell, it can be said to be defined across the (virtual) 2-D surface of the globe. I believe other components (e.g., land, ocean) use the same convention.

An alternative idea for why they are called 2-D is because until relatively recently in the history of global modeling, atmosphere models were based on rectangular (lat/lon) grids, even a 2-D field was also a 2-D Fortran array.

I have no idea if one of these is the true history of the convention to call physical fields 2-D (single level or column integral in every grid box) or 3-D (defined at multiple levels in every grid box) but I'd bet on the physics version (first explanation).